File: agent4/engine/tier1.go
package engine

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strings"
)

// PatternEntry is a single entry in the pattern hash table.
type PatternEntry struct {
	Pattern     string `json:"pattern"`
	ASLTemplate string `json:"asl_template"`
	regex       *regexp.Regexp
	slotNames   []string
	specificity int // literal word count (more = more specific)
}

// Tier1 is the pattern hash table engine.
type Tier1 struct {
	patterns []PatternEntry
}

// NewTier1 loads patterns from a JSON file and compiles them.
func NewTier1(path string) (*Tier1, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read patterns: %w", err)
	}
	var raw []PatternEntry
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("parse patterns: %w", err)
	}
	for i := range raw {
		if err := raw[i].compile(); err != nil {
			return nil, fmt.Errorf("compile pattern %q: %w", raw[i].Pattern, err)
		}
	}
	return &Tier1{patterns: raw}, nil
}

// compile converts the template pattern into a regex with named capture groups.
func (p *PatternEntry) compile() error {
	re, names := buildSlotRegex(p.Pattern)
	if re != nil {
		p.regex = re
		p.slotNames = names
	} else {
		norm := normalize(p.Pattern)
		escaped := regexp.QuoteMeta(norm)
		compiled, err := regexp.Compile("^" + escaped + "$")
		if err != nil {
			return err
		}
		p.regex = compiled
	}
	p.specificity = len(strings.Fields(p.Pattern)) - len(p.slotNames)
	return nil
}

// Translate tries all patterns and returns the most specific match
// (highest literal word count). Returns zero value and false on no match.
func (t *Tier1) Translate(sentence string) (TranslationOutput, bool) {
	norm := normalize(sentence)
	bestSpec := -1
	var bestPattern *PatternEntry
	var bestMatch []string

	for i := range t.patterns {
		p := &t.patterns[i]
		match := p.regex.FindStringSubmatch(norm)
		if match == nil {
			continue
		}
		if p.specificity > bestSpec {
			bestSpec = p.specificity
			bestPattern = p
			bestMatch = match
		}
	}

	if bestPattern == nil {
		return TranslationOutput{}, false
	}

	slots := make(map[string]string)
	for i, name := range bestPattern.regex.SubexpNames() {
		if i > 0 && name != "" {
			slots[name] = strings.ToUpper(bestMatch[i])
		}
	}

	gloss := fillASLTemplate(bestPattern.ASLTemplate, slots)
	return TranslationOutput{
		ASLGloss:   gloss,
		Method:     "pattern_hash",
		Pattern:    bestPattern.Pattern,
		Confidence: 1.0,
	}, true
}

File: agent4/asl_engine/tier1.py
from __future__ import annotations

import json
import re
from pathlib import Path

from .normalize import normalize, fill_asl_template, build_slot_regex
from .types import TranslationOutput


class PatternEntry:
    __slots__ = ("pattern", "asl_template", "regex", "slot_names", "specificity")

    def __init__(self, pattern: str, asl_template: str):
        self.pattern = pattern
        self.asl_template = asl_template
        self.regex: re.Pattern
        self.slot_names: list[str]
        self.specificity: int
        self._compile()

    def _compile(self) -> None:
        regex, names = build_slot_regex(self.pattern)
        if regex is not None:
            self.regex = regex
            self.slot_names = names
        else:
            norm = normalize(self.pattern)
            self.regex = re.compile(f"^{re.escape(norm)}$")
            self.slot_names = []
        # Specificity = literal word count (more literal = more specific)
        self.specificity = len(self.pattern.split()) - len(self.slot_names)


class Tier1:
    """Pattern hash table with template slot extraction.

    When multiple patterns match, the most specific one wins
    (highest literal word count).
    """

    def __init__(self, path: str | Path):
        data = json.loads(Path(path).read_text())
        self.patterns = [
            PatternEntry(e["pattern"], e["asl_template"]) for e in data
        ]

    def translate(self, sentence: str) -> tuple[TranslationOutput | None, bool]:
        norm = normalize(sentence)
        best: tuple[PatternEntry, re.Match] | None = None
        best_spec = -1
        for p in self.patterns:
            m = p.regex.match(norm)
            if m is None:
                continue
            if p.specificity > best_spec:
                best_spec = p.specificity
                best = (p, m)
        if best is None:
            return None, False
        p, m = best
        slots = {k: v.upper() for k, v in m.groupdict().items()}
        gloss = fill_asl_template(p.asl_template, slots)
        return TranslationOutput(
            asl_gloss=gloss,
            method="pattern_hash",
            pattern=p.pattern,
            confidence=1.0,
        ), True
